src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -28,130 +36,229 @@ public static class CachedExpressionCompiler
 
-            private static readonly ConcurrentDictionary<MemberInfo, Func<TModel, TResult>> _simpleMemberAccessCache =
-                new ConcurrentDictionary<MemberInfo, Func<TModel, TResult>>();
+            private static readonly ConcurrentDictionary<MemberInfo, Func<TModel, object>> _simpleMemberAccessCache =
+                new ConcurrentDictionary<MemberInfo, Func<TModel, object>>();
 
-            private static readonly ConcurrentDictionary<MemberInfo, Func<object, TResult>> _constMemberAccessCache =
-                new ConcurrentDictionary<MemberInfo, Func<object, TResult>>();
+            private static readonly ConcurrentDictionary<MemberInfo, Func<TModel, object>> _constMemberAccessCache =
+                new ConcurrentDictionary<MemberInfo, Func<TModel, object>>();
 
-            public static Func<TModel, TResult> Compile(Expression<Func<TModel, TResult>> expression)
+            public static Func<TModel, object> Compile(Expression<Func<TModel, TResult>> expression)
             {
-                if (expression == null)
+                switch (expression.Body)
+                {
+                    // model => model
+                    case var body when body == expression.Parameters[0]:
+                        return CompileFromIdentityFunc(expression);
+
+                    // model => (object){const}
+                    case ConstantExpression constantExpression:
+                        return CompileFromConstLookup(constantExpression);
+
+                    // model => StaticMember
+                    case MemberExpression memberExpression when memberExpression.Expression == null:
+                        return CompileFromStaticMemberAccess(expression, memberExpression);
+
+                    // model => CapturedConstant
+                    case MemberExpression memberExpression when memberExpression.Expression is ConstantExpression constantExpression:
+                        return CompileCapturedConstant(expression, memberExpression, constantExpression);
+
+                    // model => model.Property
+                    case MemberExpression memberExpression when memberExpression.Expression == expression.Parameters[0]:
+                        return CompileFromSimpleMemberAccess(expression, memberExpression);
+
+                    case MemberExpression memberExpression when IsChainedPropertyAccessor(memberExpression):
+                        return CompileForChainedMemberAccess(expression);
+
+                    default:
+                        return CompileSlow(expression);
+                }
+
+                bool IsChainedPropertyAccessor(MemberExpression memberExpression)
                 {
-                    throw new ArgumentNullException(nameof(expression));
+                    while (memberExpression.Expression != null)
+                    {
+                        if (memberExpression.Expression is MemberExpression parentExpression)
```

Comment:
Suggest `leftExpresion`. `parentExpression` is confusing here because original node (`memberExpression`) is the rightmost expression (when chain is written out) and the top of the expression tree.